### ISSUE TITLE
tokenRoot setting

### ISSUE THREAD

**Issue by simkepal** [ronzeidman/ng2-ui-auth#50]
Hello!

Great job with this. But I found a problem. My server always returns....
`{
     "error":false,
     "message":"signin-ok",
     "data":{
          "token":"*******"
     },
     "statusCode":200
}`

I also set tokenRoot in config to data (tokenRoot = 'data';)

I Always get on error, because token is not set in root response.
I think you should improve a line in setToken function to something like this:
`var accessToken = response && response.json() && (response.json().access_token || response.json().token || response.json()[this.config.tokenRoot].token);`

Thank you!

**Comment by ronzeidman** [comment]
Seems reasonable. I think I'll let you provide a configurable `resolveToken(res: Response): string` function.

**Comment by simkepal** [comment]
Great!

**Comment by ronzeidman** [comment]
Try testing the master branch (`npm i -S ronzeidman/ng2-ui-auth`) and tell me if it works for you.

**Comment by simkepal** [comment]
Hello,

I just tested this out. If I create a function in my config it is working OK. 
But now everyone with server response like mine or similar has to create this function and tokenRoot in config does not mean anything. Or am I wrong?

Thank you for your quick fix!!!!

**Comment by ronzeidman** [comment]
You're right, try now it should work for you out-of-the-box without changing the resolveToken function or any other setting (since the token is directly under "data" with name of "token")

**Comment by simkepal** [comment]
I already created a function and for me is this not a problem. Why not add something like this somewhere in function?

`var accessToken = response && response.json() && (response.json().access_token || response.json().token || response.json()[_this.tokenRoot][_this.tokenName]);`

**Comment by ronzeidman** [comment]
I did something similar, the token root should be a path and not a single property (like "prop1.prop3") so your solution is lacking. And in the Satellizer library the tokenRoot starts after the "data" property so for easy migration it should be compatible. 
If you wish you can look at the code now, I've cleaned it up a bit: https://github.com/ronzeidman/ng2-ui-auth/blob/master/src/config.service.ts#L110

**Comment by simkepal** [comment]
You are right! But if anyone has different response it will not work. If you don't find anything at the beginning you return null and the function exits. 

if (!accessToken) {
            console.warn('No token found');
            return null;
        }

I would do like this: 

`const accessToken: string | Object | null | undefined = response && response.json() &&
            (response.json().access_token || response.json().token || response.json().data);
if (!accessToken) {
    accessToken = response.json();
}
if (typeof accessToken === 'string') {
    return accessToken;
}
if (typeof accessToken !== 'object') {
    console.warn('No token found');
    return null;
}
const tokenRootData = this.tokenRoot &&
    this.tokenRoot.split('.').reduce(
        (o, x) => {
            return o[x];
        },
        accessToken);
const token = tokenRootData ? tokenRootData[this.tokenName] : accessToken[this.tokenName];
if (token) {
    return token;
}
let tokenPath = this.tokenRoot ? this.tokenRoot + '.' + this.tokenName : this.tokenName;
console.warn('Expecting a token named "' + tokenPath);
return null;`

**Comment by ronzeidman** [comment]
Maybe more like this:
```ts
const accessToken: string | Object | null | undefined = response && response.json() && 
    (response.json().access_token || response.json().token || response.json().data || response.json());
```
But then if someone has a response of 
```json
{
  "data": "some other data",
  "real-token": "actual token"
}
```
and a tokenRoot of "real-token" it won't work anyway and it would be harder to explain why...

**Comment by simkepal** [comment]
So that is why everyone should define tokenRoot and tokenName. Or not? But you are right that this error si harder to explain.

For me everything works. :)

**Comment by ronzeidman** [comment]
It would have been easier that way, but this library should work with minimal configuration for the majority of cases...

Great, I'll do some more tests of my own and publish this version to npm.

**Comment by simkepal** [comment]
Great!

Thank you for this!

**Comment by ronzeidman** [comment]
fixed in 4.1.0
